sherpa/models/tests/test_model.py

```diff
@@ -149,11 +158,77 @@ def test_complex_expression(self):
 
     def test_filter(self):
         m = self.s[::2]
-        self.assertTrue(type(m) is FilterModel)
+        self.assertIs(type(m), FilterModel)
         self.assertTrue(numpy.all(m(self.xx) == self.s(self.xx)[::2]))
 
     def test_nested(self):
         for func in (numpy.sin, self.s):
             m = self.m.apply(func)
-            self.assertTrue(type(m) is NestedModel)
+            self.assertIs(type(m), NestedModel)
             self.assertEqual(m(self.x), func(self.m(self.x)))
+
+
+# Test support for renamed parameters by sub-classing
+# the Sin model (which lets the tests be re-used).
+#
+class RenamedPars(Sin):
+
+    def __init__(self, name='renamedpars'):
+        self._renamedpars = [('norm', 'ampl')]
```

Comment:
Hmm... is this really a case we want to support? I.e. that a model might be an extension of an existing model, just defining a rename?

The approach I was hoping we could take is for `Parameter` to be aware of its *deprecated aliases*. However, this test case seems to make that approach a dead end, in that it's actually the `Model`'s responsibility to define the deprecated aliases. There is a redundancy in that being defined as instance attributes (not a great idea to begin with, imo), parameters have two names: one defined by the model, i.e. the name of the attribute, and one being the name of the parameter being passed to the parameter's initializer. So we are in a grey area where knowing the parameter's name is a responsibility of both classes.

What breaks the tie is whether we want to support this test case or not, which I don't really find too compelling, especially since we are introducing this "renaming" business ourselves, and I think that making it a parameter's responsibility to know of its aliases (which the model then needs to index, though) makes for a cleaner design.

Reply:
a) I could come up with a hand-wavy case for allowing renames, but it would not be very strong

b) this test case is - I think - in part to support some of the tests I wrote which use a renamed class to make writing the tests less onerous (switching to pytest parameterize may avoid this; I haven't thought that through)

c) how would a user/code accessing "model.deprecatedparametername" get passed on to the correct parameter? Would you have to query each parameter and ask it "do you recognize this alias"?

Reply:
a) I think it's an arbitrary decision, then. Do we want it or not?

b) which is why in my yet-to-push rework I am changing the implementation of the `RenamedPars` model to inherit the methods from `Sin` but then I short-circuited `Sin` altogether in the initializer. Rewriting the tests in `pytest` would be useful, but at this point I would leave them alone so we don't risk changing the coverage.

c) a `Parameter` has an `aliases` attribute with the deprecated aliases. The model uses it to build a dictionary name/alias -> parameter which is how `__getattr__` returns the right parameter. If the requested attribute has a name in the list of aliases for that parameter, then the warning is issued. I still have to update the xspec models to use my new stuff, but maybe I should publish it first. And by the way, tests should be failing but don't... I am wondering how much of the renamed parameters we should test in the xspec models.

Reply:
a) I suggest publishing what you have, so we can see / try it out (it would be nice to get it into the CIAO daily build for this).

b) I agree that re-writing the tests to use pytest is not for this PR (or, whatever PR gets used)

c) I feel we have to test some of the renamed parameter stuff. It may not need to be attached to the XSPEC models, since this is intended as a general purpose feature, so as long as the inheritance works okay we don't need to try every possible model combination. Or am I missing part of the problem?

Reply:
Did you get notified of my PR against your repo?

Reply:
I don't tend to look at the email address that I use gor my github account during work hours. Let me check. I got an email about PR #384, and emails about these comments (including 'Did you get notified of my PR against your repo).

Reply:
@DougBurke I issued a PR against your repo a few months ago, but I don't see it merged. Please let me know what to do. Should I just force push my rework, issue a new PR, or wait for you to review the rework the way I had pushed it?

Reply:
Wait, I think I misinterpreted my notes. It's been a while. Let me double check.

Reply:
I think I just have to close this PR in favor of #384. @DougBurke do you agree?

Reply:
I am taking it as a yes and closing this PR.

Reply:
I didn't misinterpret my notes after all. I had issued a PR against yours so you could sign off the changes, which you never did explicitly... I'll reopen this PR and push all my commits on top of it, then I'll close the other PR which is not against master.